url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPSystemManager.java

```diff
@@ -0,0 +1,1344 @@
+    /** Alarm severity for "Configured OCP link not available". */
+    private static final int    ALARM_LINK_NOT_AVAILABLE_SEV = 1;
+    /** The name of alarm "OCP Module cannot take calls". */
+    private static final String ALARM_CANNOT_TAKE_CALLS = "OCPModuleCannotTakeCalls";
+    /** Alarm ID for "OCP Module cannot take calls". */
+    private static final int    ALARM_CANNOT_TAKE_CALLS_ID = ALARM_ID_BASE + 1;
+    /** Alarm severity for "OCP Module cannot take calls". */
+    private static final int    ALARM_CANNOT_TAKE_CALLS_SEV = 3;
+
+    // Configuration
+    /** The properties object. */
+    private Properties prop;
+    /** The master/slave swap timeout in milliseconds. */
+    private int        masterSlaveSwapTimeout;
+    /** The link suspect timeout in milliseconds. */
+    private int        linkSuspectTimeout;
+
+    /** An array of the links. */
+    private CopyOnWriteArrayList<Link> links;
+    /** The system state. */
+    private SystemState     state = SystemState.STOPPED;
+    /** The preferred link. */
+    private Link            preferredLink;
+    /** The system thread. */
+    private SystemThread    sysThread;
+    /** The timestamp of when the MISSING_MASTER state was entered. */
+    private long            missingMasterStart;
+    /** True if the gapping alarm is currently active. */
+    private boolean         gappingAlarmActive = false;
+    /** The watchdog to use. */
+    private GenericWatchdog watchdog;
+
+    /** The logger for this class. */
+    private static Logger log = Logger.getLogger("ocpSystemManager");
+
+    /** The alarms interface. */
+    private static AlarmInterface alarms = UtilitiesFactory
+            .getUtilInterface().getAlarmInterface();
+
+    static {
+        alarms.registerAlarm(ALARM_LINK_NOT_AVAILABLE, Link.class,
+                ALARM_LINK_NOT_AVAILABLE_ID, ALARM_LINK_NOT_AVAILABLE_SEV);
+        alarms.registerAlarm(ALARM_CANNOT_TAKE_CALLS, OCPSystemManager.class,
+                ALARM_CANNOT_TAKE_CALLS_ID, ALARM_CANNOT_TAKE_CALLS_SEV);
+    }
+
+    /**
+     * Creates a disconnected OCP System Manager using the properties specified
+     * in <tt>jOCP.properties</tt>.
+     */
+    public OCPSystemManager() {
+        FileInputStream in = null;
+        prop = new Properties();
+
+        try {
+            in = new FileInputStream("jOCP.properties");
+            prop.load(in);
+        } catch (FileNotFoundException e) {
+            log.warn("Could not find the properties file \"jOCP.properties\". "
+                    + "Falling back to defaults.", e);
+        } catch (IOException e) {
+            log.error("Error loading properties file \"jOCP.properties\". "
+                    + "Falling back to defaults.", e);
+        } finally {
+            try {
+                if (in != null) {
+                    in.close();
+                }
+            } catch (IOException e) { // CSIGNORE: EmptyBlock
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 252:268, 427:443.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.